### PR TITLE
GH-37254: [Python] Parametrize all pickling tests to use both the pickle and cloudpickle modules

### LINKED ISSUE #37254: [Python] Include cloudpickle in pytests that pickle objects
### Describe the enhancement requested

Right now, all but one of the pytests that test pickle functionality only tests with the default python `pickle` module. Let's parameterize these tests to also test with the `cloudpickle` module.

### Component(s)

Python

### PR DESCRIPTION
**Pull request by danepitkin**
### Rationale for this change

Cloudpickle was not tested in most parts of the pyarrow test suite. Improving this coverage will make the Cython 3.0.0 upgrade cleaner as cloudpickle was failing in a few places where the default pickle module was not. This has been verified using Cython 0.29.36.

### What changes are included in this PR?

* `__reduce__` methods that need to pass kwargs have been changed from classmethod to staticmethod
* All pytests that pickle objects are parameterized to use both `pickle` and `cloudpickle`

### Are these changes tested?

Yes, pytests run successfully with Cython 0.29.36

### Are there any user-facing changes?

No.
* Closes: #37254

### CODE REVIEW

**Comment by danepitkin** [comment]
Cython 3 will set `binding=True` by default for the compiler. With Cython 0.29.36, we just need to add the binding=True compiler option to make it backwards compatible. Here's a small repro as an example:

my_test.pyx
```
# cython : language_level = 3
import cython

cdef class Foo():
    @staticmethod
    @cython.binding(True)  # This works
    def _reconstruct():
        return Foo()

    def __reduce__(self):
        return Foo._reconstruct, tuple()


@staticmethod
def _reconstruct(x):
    return Bar(x)


cdef class Bar():
    @staticmethod  # This does not work
    def _reconstruct():
        return Bar()

    def __reduce__(self):
        return Bar._reconstruct, tuple()
```

Usage here:
```
In [1]: from my_test import Foo, Bar

In [2]: f = Foo()

In [3]: b = Bar()

In [4]: import pickle

In [5]: pickle.dumps(f)
Out[5]: b'\x80\x04\x95#\x00\x00\x00\x00\x00\x00\x00\x8c\x07my_test\x94\x8c\x10Foo._reconstruct\x94\x93\x94)R\x94.'

In [6]: pickle.dumps(b)
---------------------------------------------------------------------------
PicklingError                             Traceback (most recent call last)
Cell In[6], line 1
----> 1 pickle.dumps(b)

PicklingError: Can't pickle <built-in function _reconstruct>: it's not the same object as my_test._reconstruct
```

**Comment by danepitkin** [comment]
There seems to be an intermittent failure reported:

```
==================================== ERRORS ====================================
___ ERROR at setup of test_filesystem_pickling[builtin_pickle-S3FileSystem] ____

request = <SubRequest 's3fs' for <Function test_filesystem_pickling[builtin_pickle-S3FileSystem]>>
s3_server = {'connection': ('localhost', 38507, 'arrow', 'apachearrow'), 'process': <Popen: returncode: None args: ['minio', '--compat', 'server', '--quiet', '-...>, 'tempdir': '/tmp/tmpoas4aj3y'}

    @pytest.fixture
    def s3fs(request, s3_server):
        request.config.pyarrow.requires('s3')
        from pyarrow.fs import S3FileSystem
    
        host, port, access_key, secret_key = s3_server['connection']
        bucket = 'pyarrow-filesystem/'
    
        fs = S3FileSystem(
            access_key=access_key,
            secret_key=secret_key,
            endpoint_override='{}:{}'.format(host, port),
            scheme='http',
            allow_bucket_creation=True,
            allow_bucket_deletion=True
        )
>       fs.create_dir(bucket)

opt/conda/envs/arrow/lib/python3.9/site-packages/pyarrow/tests/test_fs.py:250: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pyarrow/_fs.pyx:603: in pyarrow._fs.FileSystem.create_dir
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   OSError: When creating bucket 'pyarrow-filesystem': AWS Error UNKNOWN (HTTP status 503) during HeadBucket operation: No response body.
```

File: python/pyarrow/_dataset_parquet.pyx [hunk 1]
```diff
@@ -770,9 +771,12 @@ cdef class ParquetFragmentScanOptions(FragmentScanOptions):
             other.thrift_container_size_limit)
         return attrs == other_attrs
 
-    @classmethod
-    def _reconstruct(cls, kwargs):
-        return cls(**kwargs)
+    @staticmethod
+    @binding(True)  # Required for cython < 3
```
Comment (kou) [inline python/pyarrow/_dataset_parquet.pyx:775]:
```suggestion
    @binding(True)  # Required for Cython < 3
```

File: python/pyarrow/tests/parquet/test_dataset.py [hunk 2]
```diff
@@ -1557,20 +1557,10 @@ def is_pickleable(obj):
 
 @pytest.mark.pandas
 @parametrize_legacy_dataset
-def test_builtin_pickle_dataset(tempdir, datadir, use_legacy_dataset):
-    import pickle
+def test_pickle_dataset(tempdir, datadir, use_legacy_dataset, pickle_module):
     dataset = _make_dataset_for_pickling(tempdir, use_legacy_dataset)
     _assert_dataset_is_picklable(
```
Comment (kou) [inline python/pyarrow/tests/parquet/test_dataset.py:1562]: Can we inline `_assert_dataset_is_picklable()` here?

**Review by kou — COMMENTED** [review]
(Sorry, I can't review this deeply because I'm not familiar with pickle.)

File: python/pyarrow/tests/parquet/test_dataset.py (hunk 2 above)
Comment (danepitkin) [reply python/pyarrow/tests/parquet/test_dataset.py:1562]: Will update! Good idea.

**Comment by danepitkin** [comment]
For the intermittent failure, I've also seen it in the CI tests for https://github.com/apache/arrow/pull/37097. The `503` error means "Service unavailable". Seeing as this is a session-scoped pytest fixture and it only is failing on the very first invocation so far, I'm pretty sure that minio is just not fully started yet when the pytest tries to call the server after starting it. We can introduce pytest_xprocess to better manage processes as fixtures. I'll add that to this PR. https://pytest-xprocess.readthedocs.io/en/latest/

**Comment by danepitkin** [comment]
> For the intermittent failure, I've also seen it in the CI tests for #37097. The `503` error means "Service unavailable". Seeing as this is a session-scoped pytest fixture and it only is failing on the very first invocation so far, I'm pretty sure that minio is just not fully started yet when the pytest tries to call the server after starting it the first time. We can introduce pytest_xprocess to better manage processes as fixtures. I'll add that to this PR. https://pytest-xprocess.readthedocs.io/en/latest/

xprocess doesn't quite fit cleanly into what we want. I tried adding a connection_timeout parameter to s3fs in our tests instead to see if that helps.

**Comment by danepitkin** [comment]
@github-actions crossbow submit python

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
@github-actions crossbow submit test-conda-python-3.11-hypothesis

**Comment by danepitkin** [comment]
Still need to fix the intermittent failure with minio server startup.

**Review by jorisvandenbossche — APPROVED** [review]
Actual code changes (apart from the S3 issue) look good, thanks a lot for this!

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
@github-actions crossbow submit *python*

**Comment by danepitkin** [comment]
CI looks good! No S3 intermittent failures reported after 2 full runs. All current failures are unrelated.

Ready to merge IMO.

**Comment by jorisvandenbossche** [comment]
Nice, thanks a lot Dane!

**Pull request merged**